timemachine/fe/geometry.py

```diff
@@ -3,56 +3,106 @@
-        if hybridization == HybridizationType.SP3:
-            return LocalGeometry.G3_PYRAMIDAL
-        elif hybridization == HybridizationType.SP2:
-            return LocalGeometry.G3_PLANAR
-        else:
-            assert 0, "Unknown 3-nbr geometry"
-    elif len(nbrs) == 4:
-        if hybridization == HybridizationType.SP3:
-            return LocalGeometry.G4_TETRAHEDRAL
-        else:
-            assert 0, "Unknown 4-nbr geometry"
-    else:
-        assert 0, "Too many neighbors"
-
-
-def classify_geometry(mol: Chem.Mol) -> List[LocalGeometry]:
+    cmat = np.zeros((num_atoms, num_atoms))
+    for i, j in bond_idxs:
+        cmat[i][j] = 1
+        cmat[j][i] = 1
+
+    nblist = []
+    for i in range(num_atoms):
+        nbs = []
+        for j in range(num_atoms):
+            if cmat[i][j]:
+                nbs.append(j)
+        nblist.append(nbs)
+
+    return nblist
+
+
+def label_stereo(
+    num_atoms,
+    bond_idxs,
+    angle_idxs,
+    angle_params,
+    improper_idxs,
+    improper_params,
+):
+    # list of list representation
+    nblist = bond_idxs_to_nblist(num_atoms, bond_idxs)
+    atom_geometries = []
+    for atom_idx, atom_nbs in enumerate(nblist):
+        if len(atom_nbs) == 4:
+            atom_geometries.append(LocalGeometry.G4_TETRAHEDRAL)
+        elif len(atom_nbs) == 3:
+            # check for impropers
+            local_geometry = LocalGeometry.G3_PYRAMIDAL
+            # (tbd): should we also check propers as well?
+            # impropers are centered around the first atom.
+            for (i, _, _, _), (k, p, n) in zip(improper_idxs, improper_params):
+                if i == atom_idx and k > 0.0:
+                    local_geometry = LocalGeometry.G3_PLANAR
+                    break
+
+            atom_geometries.append(local_geometry)
+            # atom_stereo_flags.append(is_stereo)
+        elif len(atom_nbs) == 2:
+            # check angle terms:
+            local_geometry = LocalGeometry.G2_KINK
+            for (i, j, k), (_, angle) in zip(angle_idxs, angle_params):
+                if abs(angle) < 0.05:
+                    assert 0
+                ii, kk = atom_nbs[0], atom_nbs[1]
+                if j == atom_idx:
+                    if (i, k) == (ii, kk) or (i, k) == (kk, ii):
+                        if abs(angle - np.pi) < 0.05:
+                            local_geometry = LocalGeometry.G2_LINEAR
+                            break
+            atom_geometries.append(local_geometry)
+        elif len(atom_nbs) == 1:
+            atom_geometries.append(LocalGeometry.G1_TERMINAL)
+        elif len(atom_nbs) == 0:
+            atom_geometries.append(None)
+
+    return atom_geometries
+
+
+def classify_geometry(mol: Chem.Mol, ff: Forcefield = None, core: List[int] = None) -> List[LocalGeometry]:
     """
     Identify the local geometry of the molecule. This current uses a heuristic but we
     should really be generating this from gas-phase simulations of the real forcefield.
```

Comment:
Is this comment still valid? I.e. we should to run gas-phase simulations then get the geometry?

Reply:
We should eventually classify geometries by running simulations. Right now we simply inspect the forcefield parameters pray to the stat mech gods.